draft-ietf-tls-external-psk-importer.md

```diff
@@ -99,53 +101,68 @@ a tuple of (Base Key, External Identity, KDF). The associated KDF (and hash func
 - External Identity: The identity of an EPSK.
 - Imported Identity: The identity of a PSK as sent on the wire.
 
+## Notation and Cryptographic Dependencies
+
+The key importer API depends on a generic KDF which provides the following interface.
+
+- Nh: The output size of the Extract function.
+- Extract(salt, IKM): Extract a pseudorandom key of fixed length Nh
+  from input keying material `IKM` and an optional octet string
+  `salt`.
+- Expand(PRK, info, L): Expand a pseudorandom key `PRK` using
+  optional string `info` into `L` bytes of output keying material.
+
+It also depends on the following utility function.
+
+- `concat(x0, ..., xN)`: Concatenation of octet strings.
+  `concat(0x01, 0x0203, 0x040506) = 0x010203040506`
+
 # Key Import
 
 A key importer takes as input an EPSK with external identity 'external_identity' and base key 'epsk',
-as defined in {{terminology}}, along with an optional label, and transforms
-it into a set of PSKs and imported identities for use in a connection based on supported HashAlgorithms.
-In particular, for each supported HashAlgorithm 'hash', the importer constructs an ImportedIdentity
-structure as follows:
+as defined in {{terminology}}, along with an optional context string, and transforms
+it into a set of PSKs and imported identities for use in a connection based on the (D)TLS protocol
+version and a specific KDF. In particular, for each supported protocol version 'protocol' and KDF
+function 'kdf', the importer constructs an ImportedIdentity structure as follows:
 
 ~~~
    struct {
        opaque external_identity<1...2^16-1>;
-       opaque label<0..2^8-1>;
        opaque context<0..2^16-1>;
-       HashAlgorithm hash;
+       uint16 protocol;
+       uint16 kdf;
    } ImportedIdentity;
 ~~~
 
-[[TODO: An alternative design might combine label and hash into the same field so that future
-protocols which don't have a notion of HashAlgorithm don't need this field.]]
+ImportedIdentity.context MUST include the context used to derive the EPSK, if any exists.  If the EPSK is a key derived
+from some other protocol or sequence of protocols, ImportedIdentity.context MUST include a channel binding for the deriving protocols
+{{!RFC5056}}.  If any secrets are agreed in earlier protocols they SHOULD be included in ImportedIdentity.context {{CCB}}.
```

Comment:
Shouldn't secrets be folded into the PSK?

Reply:
As far as I know, and @jhoyla should feel free to correct me, this is the "usual" (for some definition of usual) channel binding mechanism. That is, add a slot to put in additional channel binding blobs. In principle, I agree that it could (should?) be included in the external PSK, as that seems morally superior, but I defer to the channel binding expert. :-)

Reply:
So this is based on the contributive channel bindings from [BDP15](https://www.ndss-symposium.org/wp-content/uploads/2017/09/08_5.pdf). Basically for the authentication properties of outer layers to propagate inwards each layer has to prove it still has access to the previously established secrets. So yes, they should already be folded into the PSK, but if the PSK is generated by a long chain of protocols this recommits the client and server to all those sessions individually. 
Imagine that there were some attack on the layer immediately preceding the PSK importer that leaked the EPSK. By re-committing to the previously established keys even knowing the EPSK is insufficient for the attacker to mount an attack.